### ISSUE TITLE
DPI-1047: 64-bit Oracle Client library cannot be loaded

### ISSUE THREAD

**Issue by wrdaigle** [oracle/python-cx_Oracle#84]
I'm beating my head against a wall on this one.  Here is the synopsis.  I have an ArcGIS server running on a Linux box. The ArcGIS server software install include a custom version of WINE, with python running under wine.  We are tyring to run cx_Oracle within the version of Python running under WINE.  We got this working in a previous version ArcGIS server, bu so far no luck with the latest version.

#### Answer the following questions:

1. What is your version of Python? Is it 32-bit or 64-bit?
Python 2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:53:40) [MSC v.1500 64 bit (AMD64)] on win32
-running under WINE v1.9.2-1051final

2. What is your version of cx_Oracle?
6.0.2

3. What is your version of the Oracle client (e.g. Instant Client)?  How was it
   installed?  Where is it installed?
11.2.0.4 instant client - 64-bit windows version

4. What is your version of the Oracle Database?
11.2.0.4

5. What is your OS and version?
Ubuntu 16.04.3 running 64-bit version of WINE

6. What compiler version did you use?  For example, with GCC, run
   `gcc --version`.
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609 

7. What environment variables did you set?  How *exactly* did you set them?
wine regedit
HKLM>SYSTEM.CurrentControlSet>Control>Session Manager>Environment
 - added oracle instant client path to the PATH variable as I would on a windows machine

8. What *exact* command caused the problem (e.g. what command did you try to
   install with)?  Who were you logged in as?
The lbrary imports fine. 
When I tried to create a connect:
  ` c = cx_Oracle.connect('tnsname','un','pw')`
or try to check the client version:
`  cx_Oracle.clientversion()`
I get the following error:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
cx_Oracle.DatabaseError: DPI-1047: 64-bit Oracle Client library cannot be loaded: "DLL initialization failed". See https://oracle.github.io/odpi/doc/installation.html#windows for help

**Comment by cjbj** [comment]
I imagine you still need the VS Redistributable in this environment? Per the [Instant Client download page ](http://www.oracle.com/technetwork/topics/winx64soft-089540.html), Oracle 11.2 needs the Microsoft Visual Studio 2005 Redistributable.

**Comment by wrdaigle** [comment]
I have tried installing it under wine without success.  However, I'm not sure this is the problem since I can successfully use the client via sqlplus (running under WINE).

**Comment by cjbj** [comment]
If you can run SQL*Plus, you it seems logical that you'd be able to invoke other OCI programs like cx_Oracle - but you are definitely in unsupported explorer territory.

**Comment by wrdaigle** [comment]
Indeed.  I was really hoping to find a solution to this.  Since we had it working in an earlier version, I'm facing a relatively hefty rewrite in order to maintain the same functionality

**Comment by cjbj** [comment]
It would be interesting to know if cx_Oracle 5.3 works: http://cx-oracle.readthedocs.io/en/v5.x/installation.html

**Comment by wrdaigle** [comment]
I wish I could tell you. The previous version of CX Oracle we were using was 5.2 something,a nd that worked fine. I tried 5.3, but it has a requirement for visual C++ for python, and for the life of me I couldn't get that to install on wine.
________________________________________

**Comment by cjbj** [comment]
Did you try the Windows EXE from https://pypi.python.org/pypi/cx_Oracle/5.3 ?

Can you clarify whether you have any cx_Oracle version running with your current ArcGIS setup?

Triple check your PATH etc.

**Comment by wrdaigle** [comment]
Good thought. I just tried the exe. It seems to have installed successfully, but I try to import it, I get a “DLL initialization failed.” error.

All version of cx_Oracle work fine with my desktop installs ArcGIS.  The problem only exist once we get to server.

I think the root of the problem is the lack of the .NET framework with WINE on our latest server.  The previous version had several versions included (4.0, 3.0, 2.0 and 1.1) packaged with wine.  We are working on getting .NET onto the our new version to see if that makes a difference.

**Comment by cjbj** [comment]
It's still puzzling that SQL*Plus works but cx_Oracle doesn't - they rely on pretty much the same things.

**Comment by wrdaigle** [comment]
Ultimately, we end up finding a solution.  It's not pretty, but it works.  Instead of installing anything within wine itself, we installed everything on dedicated 64-bit Windows desktop and copied the compiled files over to the WINE installation on the linux server.  

We copied the following to the site-packages directory on the ArcGIS server:
-cx_oracle.pyd
-the oracle instant client dlls
Placing the windows dlls in this directory allowed cx_Oracle to function without updating the path variables associated with the linux version of the client (which is used by other aspects of the ArcGIS software)

Additionally, we copied the Microsoft Framework folder (C:\Windows\Microsoft.NET\Framework).

With this method, we were able to get the 5.x versions of cx_Oracle to function.  Still no luck with 6.x, but we can live without that for now.  

Thanks for the help!
